File: skills/udf-gen-test/templates/java/src/main/java/com/udf/bench/GenData.java
/*
 * SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package com.udf.bench;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.udf.SparkUtils;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;

/**
 * Generates benchmark data and optionally validates by running
 * BenchUtils.executeCpu and BenchUtils.executeGpu.
 *
 * Usage:
 *   mvn exec:java -Dexec.mainClass=com.udf.bench.GenData \
 *     -Dexec.args="--rows 1000 --validate --spark-conf k=v ..."
 */
public class GenData {

    public static void main(String[] args) {
        Map<String, String> argMap = new HashMap<>();
        List<String> sparkConfs = new ArrayList<>();
        parseArgs(args, argMap, sparkConfs);

        long rows = Long.parseLong(SparkUtils.requireArg(argMap, "rows"));
        int partitions = Integer.parseInt(argMap.getOrDefault("partitions", "32"));
        boolean validate = argMap.containsKey("validate");
        String outputPath = argMap.get("output-path");

        // Build Spark session
        SparkSession.Builder builder = SparkSession.builder().appName("GenData");
        SparkUtils.applySparkConfs(builder, sparkConfs);
        SparkSession spark = builder.enableHiveSupport().getOrCreate();

        try {
            // Generate synthetic data
            Dataset<Row> df = BenchUtils.generateSyntheticData(spark, rows, partitions);

            // Verify row count
            long actualRows = df.count();
            if (actualRows != rows) {
                System.err.println("Row count mismatch: expected=" + rows
                    + ", actual=" + actualRows);
                System.exit(1);
            }
            System.out.println("Generated " + actualRows + " rows across "
                + partitions + " partitions");

            if (validate) {
                // Validation mode — run both CPU and GPU execute, don't write
                for (String label : new String[]{"cpu", "gpu"}) {
                    try {
                        if ("cpu".equals(label)) {
                            BenchUtils.executeCpu(spark, df).collect();
                        } else {
                            BenchUtils.executeGpu(spark, df).collect();
                        }
                        System.out.println("Validation (" + label + ") passed.");
                    } catch (Exception e) {
                        System.err.println("Validation (" + label + ") failed: "
                            + e.getClass().getSimpleName() + ": " + e.getMessage());
                        e.printStackTrace(System.err);
                        System.exit(1);
                    }
                }
            } else {
                // Generation mode — write to output path
                if (outputPath == null) {
                    throw new IllegalArgumentException(
                        "--output-path is required when not in validation mode");
                }
                df.write().mode("overwrite").parquet(outputPath);
                System.err.println("Successfully generated dataset and saved to: " + outputPath);
            }
        } catch (Exception e) {
            System.err.println("Failed to generate dataset: "
                + e.getClass().getSimpleName());
            e.printStackTrace(System.err);
            System.exit(1);
        } finally {
            spark.stop();
        }

        System.exit(0);
    }

    /** Parse CLI arguments. */
    private static void parseArgs(String[] args, Map<String, String> map, List<String> sparkConfs) {
        int i = 0;
        while (i < args.length) {
            switch (args[i]) {
                case "--rows":        map.put("rows", args[i + 1]); i += 2; break;
                case "--partitions":  map.put("partitions", args[i + 1]); i += 2; break;
                case "--validate":    map.put("validate", "true"); i += 1; break;
                case "--output-path": map.put("output-path", args[i + 1]); i += 2; break;
                case "--spark-conf":  sparkConfs.add(args[i + 1]); i += 2; break;
                default:
                    throw new IllegalArgumentException("Unknown argument: " + args[i]);
            }
        }
    }
}

File: skills/udf-gen-test/templates/scala/src/main/scala/com/udf/bench/GenData.scala
/*
 * SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package com.udf.bench

import com.udf.SparkUtils
import org.apache.spark.sql.SparkSession

/**
 * Generates benchmark data and optionally validates by running
 * BenchUtils.executeCpu and BenchUtils.executeGpu.
 *
 * Usage:
 *   mvn exec:java -Dexec.mainClass=com.udf.bench.GenData \
 *     -Dexec.args="--rows 1000 --validate --spark-conf k=v ..."
 */
object GenData {

  def main(args: Array[String]): Unit = {
    val (parsed, sparkConfs) = parseArgs(args)

    val rows = parsed.getOrElse("rows",
      throw new IllegalArgumentException("--rows is required")).toLong
    val partitions = parsed.getOrElse("partitions", "32").toInt
    val validate = parsed.contains("validate")
    val outputPath = parsed.get("output-path")

    // Build Spark session
    val builder = SparkSession.builder().appName("GenData")
    SparkUtils.applySparkConfs(builder, sparkConfs)
    val spark = builder.getOrCreate()

    try {
      // Generate synthetic data
      val df = BenchUtils.generateSyntheticData(spark, rows, partitions)

      // Verify row count
      val actualRows = df.count()
      if (actualRows != rows) {
        System.err.println(s"Row count mismatch: expected=$rows, actual=$actualRows")
        sys.exit(1)
      }
      println(s"Generated $actualRows rows across $partitions partitions")

      if (validate) {
        // Validation mode — run both CPU and GPU execute, don't write
        for ((label, executeFn) <- Seq(
          ("cpu", BenchUtils.executeCpu _),
          ("gpu", BenchUtils.executeGpu _)
        )) {
          try {
            executeFn(spark, df).collect()
            println(s"Validation ($label) passed.")
          } catch {
            case e: Exception =>
              System.err.println(
                s"Validation ($label) failed: ${e.getClass.getSimpleName}: ${e.getMessage}")
              e.printStackTrace(System.err)
              sys.exit(1)
          }
        }
      } else {
        // Generation mode — write to output path
        val path = outputPath.getOrElse(
          throw new IllegalArgumentException("--output-path is required when not in validation mode"))
        df.write.mode("overwrite").parquet(path)
        System.err.println(s"Successfully generated dataset and saved to: $path")
      }
    } catch {
      case e: Exception =>
        System.err.println(s"Failed to generate dataset: ${e.getClass.getSimpleName}")
        e.printStackTrace(System.err)
        sys.exit(1)
    } finally {
      spark.stop()
    }

    sys.exit(0)
  }

  /** Parse CLI arguments. */
  private def parseArgs(args: Array[String]): (Map[String, String], Seq[String]) = {
    var map = Map.empty[String, String]
    var sparkConfs = Seq.empty[String]
    var i = 0
    while (i < args.length) {
      args(i) match {
        case "--rows"        => map += ("rows" -> args(i + 1)); i += 2
        case "--partitions"  => map += ("partitions" -> args(i + 1)); i += 2
        case "--validate"    => map += ("validate" -> "true"); i += 1
        case "--output-path" => map += ("output-path" -> args(i + 1)); i += 2
        case "--spark-conf"  => sparkConfs :+= args(i + 1); i += 2
        case other =>
          throw new IllegalArgumentException(s"Unknown argument: $other")
      }
    }
    (map, sparkConfs)
  }
}
